### PR TITLE
[ARCTIC-257][Flink] Flink source project pushdown would cause reading all fields to downstream

### LINKED ISSUE #257: [Bug]: flink source project pushdown would cause reading all fields to downstream
### What happened?

Select partial fields from arctic table, but there will be all fields without projecting. It will cause unmatched rowtype and real rowdata. And shuffle is using all fields RowType, which may cause ArrayIndexOutOfBoundsException.

### Affects Versions

master/0.3.1

### What engines are you seeing the problem on?

Flink

### How to reproduce

Create a table t with fields a string, b int, c string.  And declare c as partition key. Then select a, c from t. 

### Relevant log output

```shell
java.lang.RuntimeException: 4
	at org.apache.flink.streaming.runtime.io.RecordWriterOutput.pushToRecordWriter(RecordWriterOutput.java:109)
	at org.apache.flink.streaming.runtime.io.RecordWriterOutput.collect(RecordWriterOutput.java:93)
	at org.apache.flink.streaming.runtime.io.RecordWriterOutput.collect(RecordWriterOutput.java:44)
	at org.apache.flink.streaming.runtime.tasks.SourceOperatorStreamTask$AsyncDataOutputToOutput.emitRecord(SourceOperatorStreamTask.java:188)
	at org.apache.flink.streaming.api.operators.source.SourceOutputWithWatermarks.collect(SourceOutputWithWatermarks.java:110)
	at org.apache.flink.streaming.api.operators.source.SourceOutputWithWatermarks.collect(SourceOutputWithWatermarks.java:101)
	at com.netease.arctic.flink.read.hybrid.reader.ArcticRecordEmitter.emitRecord(ArcticRecordEmitter.java:35)
	at com.netease.arctic.flink.read.hybrid.reader.ArcticRecordEmitter.emitRecord(ArcticRecordEmitter.java:28)
	at org.apache.flink.connector.base.source.reader.SourceReaderBase.pollNext(SourceReaderBase.java:128)
	at org.apache.flink.streaming.api.operators.SourceOperator.emitNext(SourceOperator.java:275)
	at org.apache.flink.streaming.runtime.io.StreamTaskSourceInput.emitNext(StreamTaskSourceInput.java:67)
	at org.apache.flink.streaming.runtime.io.StreamOneInputProcessor.processInput(StreamOneInputProcessor.java:65)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.processInput(StreamTask.java:398)
	at org.apache.flink.streaming.runtime.tasks.mailbox.MailboxProcessor.runMailboxLoop(MailboxProcessor.java:191)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.runMailboxLoop(StreamTask.java:619)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.invoke(StreamTask.java:583)
	at org.apache.flink.runtime.taskmanager.Task.doRun(Task.java:758)
	at org.apache.flink.runtime.taskmanager.Task.run(Task.java:573)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.ArrayIndexOutOfBoundsException: 4
	at org.apache.flink.table.data.GenericRowData.isNullAt(GenericRowData.java:129)
	at org.apache.iceberg.flink.RowDataWrapper.get(RowDataWrapper.java:68)
	at org.apache.iceberg.Accessors$PositionAccessor.get(Accessors.java:71)
	at org.apache.iceberg.Accessors$PositionAccessor.get(Accessors.java:58)
	at org.apache.iceberg.PartitionKey.partition(PartitionKey.java:104)
	at com.netease.arctic.flink.shuffle.ShuffleHelper.hashPartitionValue(ShuffleHelper.java:110)
	at com.netease.arctic.flink.shuffle.ReadShuffleRulePolicy$RoundRobinPartitioner.partition(ReadShuffleRulePolicy.java:111)
	at com.netease.arctic.flink.shuffle.ReadShuffleRulePolicy$RoundRobinPartitioner.partition(ReadShuffleRulePolicy.java:83)
	at org.apache.flink.streaming.runtime.partitioner.CustomPartitionerWrapper.selectChannel(CustomPartitionerWrapper.java:54)
	at org.apache.flink.streaming.runtime.partitioner.CustomPartitionerWrapper.selectChannel(CustomPartitionerWrapper.java:33)
	at org.apache.flink.runtime.io.network.api.writer.ChannelSelectorRecordWriter.emit(ChannelSelectorRecordWriter.java:54)
	at org.apache.flink.streaming.runtime.io.RecordWriterOutput.pushToRecordWriter(RecordWriterOutput.java:107)
	... 18 more
```

### Anything else

In #108 supports projection push down. 
But the flinkSchemaRowType in ArcticDynamicSource won't be adjust after projection, which help to build ShuffleHelper.
On the other hand, there is still using all fields schema in buiding RowDataReaderFunction in FlinkSource.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by SteNicholas**
## Why are the changes needed?

Select partial fields from arctic table, but there will be all fields without projecting. It will cause unmatched rowtype and real rowdata. And shuffle is using all fields `RowType`, which may cause `ArrayIndexOutOfBoundsException`. 

Fix #257.

## Brief change log

  - Update `readSchema` and `flinkSchemaRowType` with the projection fields for `SupportsProjectionPushDown#applyProjection`.
  - Use `scanContext#project` for the projected schema of `RowDataReaderFunction`.

## How was this patch tested?
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] Run test locally before making a pull request

## Documentation

  - Does this pull request introduces a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink/v1.12/flink/src/main/java/com/netease/arctic/flink/table/ArcticDynamicSource.java [hunk 1]
```diff
@@ -235,6 +238,16 @@ public boolean supportsNestedProjection() {
 
   @Override
   public void applyProjection(int[][] projectedFields) {
```
Comment (zstraw) [inline flink/v1.12/flink/src/main/java/com/netease/arctic/flink/table/ArcticDynamicSource.java:240]: The projection logic here should also apply to com.netease.arctic.flink.table.ArcticFileSource

**Review by zstraw — COMMENTED** [review]
Can you add some UTs for this case?

File: flink/v1.12/flink/src/main/java/com/netease/arctic/flink/table/FlinkSource.java [hunk 2]
```diff
@@ -151,7 +152,9 @@ public DataStream<RowData> build() {
       RowDataReaderFunction rowDataReaderFunction = new RowDataReaderFunction(
           flinkConf,
           arcticTable.schema(),
-          arcticTable.schema(),
+          projectedSchema == null ?
```
Comment (zstraw) [inline flink/v1.12/flink/src/main/java/com/netease/arctic/flink/table/FlinkSource.java:155]: Line 144 has checked projectedSchema==null, scanContext.project() can be used here directly.

**Comment by zstraw** [comment]
> Can you add some UTs for this case?

There are some cases in my mind.

1.  The type of fields are different in the latest two fields, and one is partition key. Select the latest one field. 
e.g. Create a table t with fields a string, b int, c string. And declare c as partition key. Then select a, c from t.

2. declare a watermark in arctic table by sql like.  And select partial fields in different orders with table schema.
e.g. create table t (ts as localtimestamp, watermark for ts as ts) like arctic_catalog.db.table

3.  dim join arctic table. And select partial fields in different orders with table schema.
e.g. create table user_dim  (
    optime TIMESTAMP(3),
    WATERMARK FOR optime AS optime 
) LIKE arctic_catalog.db.table;

**Review by zstraw — APPROVED** [review]
LGTM. The existing UTs have covered the cases.

**Review by zstraw — APPROVED** [review]

**Review by zhoujinsong — APPROVED** [review]
LGTM.

**Pull request merged**